contracts/contracts/ValidatorAuction.sol

```diff
@@ -148,19 +148,32 @@ contract ValidatorAuction is Ownable {
         require(auctionState == AuctionState.Ended || auctionState == AuctionState.Failed, "You cannot withdraw before the auction is ended or it failed.");
         require(bids[msg.sender] >= lowestBidPrice, "The sender has nothing to withdraw.");
 
-        uint valueToWithdraw;
-
         if (auctionState == AuctionState.Ended) {
-            valueToWithdraw = bids[msg.sender] - lowestBidPrice;
-            assert(valueToWithdraw <= bids[msg.sender]);
-
-            bids[msg.sender] = lowestBidPrice;
+            withdrawAfterAuctionEnded();
         } else {
-            valueToWithdraw = bids[msg.sender];
-            assert(valueToWithdraw <= bids[msg.sender]);
-
-            bids[msg.sender] = 0;
+            withdrawAfterAuctionFailed();
         }
+    }
+
+    function withdrawAfterAuctionEnded() internal stateIs(AuctionState.Ended) {
+        require(bids[msg.sender] > lowestBidPrice, "The sender has nothing to withdraw.");
+
+        uint valueToWithdraw = bids[msg.sender] - lowestBidPrice;
+        assert(valueToWithdraw <= bids[msg.sender]);
+
+        bids[msg.sender] = lowestBidPrice;
+
+        msg.sender.transfer(valueToWithdraw);
+    }
+
+    function withdrawAfterAuctionFailed() internal stateIs(AuctionState.Failed) {
+        require(bids[msg.sender] >= 0, "The sender has nothing to withdraw.");
```

Comment:
Here you could no withdraw twice. I would make it symmetric  to the auctionEnded case and use > 0.